File: elixir/lib/scratchbird/scram.ex
# ScratchBird-driver
# Copyright (c) 2025-2026 Dalton Calford
#
# Licensed under the Initial Developer's Public License Version 1.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at:
# https://www.firebirdsql.org/en/initial-developer-s-public-license-version-1-0/

defmodule ScratchBird.Scram do
  @moduledoc false

  defstruct [:username, :client_nonce, :client_first_bare, :server_signature]

  def new(username) do
    nonce = :crypto.strong_rand_bytes(18) |> Base.encode64()
    %__MODULE__{username: username, client_nonce: nonce}
  end

  def client_first(%__MODULE__{username: username, client_nonce: nonce} = state) do
    bare = "n=#{escape(username)},r=#{nonce}"
    {"n,," <> bare, %{state | client_first_bare: bare}}
  end

  def handle_server_first(%__MODULE__{} = state, password, server_first) do
    attrs = parse_attrs(server_first)
    nonce = Map.get(attrs, "r", "")
    salt = Map.get(attrs, "s", "")
    iter = Map.get(attrs, "i", "0")

    if nonce == "" or not String.starts_with?(nonce, state.client_nonce) do
      {:error, "SCRAM server nonce mismatch"}
    else
      iterations = String.to_integer(iter)
      salt_bin = Base.decode64!(salt)
      salted = pbkdf2_sha256(password, salt_bin, iterations, 32)
      client_key = hmac_sha256(salted, "Client Key")
      stored_key = :crypto.hash(:sha256, client_key)
      client_final = "c=biws,r=" <> nonce
      auth_message = state.client_first_bare <> "," <> server_first <> "," <> client_final
      client_sig = hmac_sha256(stored_key, auth_message)
      client_proof = xor_bytes(client_key, client_sig)
      server_key = hmac_sha256(salted, "Server Key")
      server_sig = hmac_sha256(server_key, auth_message)
      response = client_final <> ",p=" <> Base.encode64(client_proof)
      {:ok, response, %{state | server_signature: server_sig}}
    end
  end

  def verify_server_final(%__MODULE__{server_signature: sig}, server_final) do
    attrs = parse_attrs(server_final)
    verifier = Map.get(attrs, "v")
    expected = Base.encode64(sig || <<>>)
    if verifier == expected, do: :ok, else: {:error, "SCRAM server signature mismatch"}
  end

  defp escape(text) do
    text
    |> String.replace("=", "=3D")
    |> String.replace(",", "=2C")
  end

  defp parse_attrs(message) do
    message
    |> String.split(",", trim: true)
    |> Enum.reduce(%{}, fn part, acc ->
      case String.split(part, "=", parts: 2) do
        [k, v] -> Map.put(acc, k, v)
        _ -> acc
      end
    end)
  end

  defp hmac_sha256(key, data) when is_binary(data) do
    :crypto.mac(:hmac, :sha256, key, data)
  end

  defp hmac_sha256(key, data) when is_list(data) do
    :crypto.mac(:hmac, :sha256, key, to_string(data))
  end

  defp xor_bytes(left, right) do
    left
    |> :binary.bin_to_list()
    |> Enum.zip(:binary.bin_to_list(right))
    |> Enum.map(fn {l, r} -> Bitwise.bxor(l, r) end)
    |> :binary.list_to_bin()
  end

  defp pbkdf2_sha256(password, salt, iterations, key_len) do
    :crypto.pbkdf2_hmac(:sha256, password, salt, iterations, key_len)
  end
end

File: tracks/p3/drivers/rust/src/scram.rs
// ScratchBird-driver
// Copyright (c) 2025-2026 Dalton Calford
//
// Licensed under the Initial Developer's Public License Version 1.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at:
// https://www.firebirdsql.org/en/initial-developer-s-public-license-version-1-0/
use base64::{engine::general_purpose, Engine as _};
use hmac::{Hmac, Mac};
use pbkdf2::pbkdf2_hmac;
use rand::RngCore;
use sha2::Sha256;

use crate::errors::{Error, ErrorKind, Result};

type HmacSha256 = Hmac<Sha256>;

pub struct ScramExchange {
    username: String,
    client_nonce: String,
    client_first_bare: String,
    server_signature: Option<Vec<u8>>,
}

impl ScramExchange {
    pub fn new(username: &str) -> Self {
        let mut nonce = [0u8; 18];
        rand::thread_rng().fill_bytes(&mut nonce);
        Self {
            username: username.to_string(),
            client_nonce: general_purpose::STANDARD.encode(nonce),
            client_first_bare: String::new(),
            server_signature: None,
        }
    }

    pub fn client_first_message(&mut self) -> String {
        let escaped = self.username.replace("=", "=3D").replace(",", "=2C");
        self.client_first_bare = format!("n={},r={}", escaped, self.client_nonce);
        format!("n,,{}", self.client_first_bare)
    }

    pub fn handle_server_first(&mut self, password: &str, server_first: &str) -> Result<String> {
        let attrs = parse_attributes(server_first);
        let nonce = attrs
            .get("r")
            .ok_or_else(|| Error::new(ErrorKind::Auth, "SCRAM missing nonce"))?;
        if !nonce.starts_with(&self.client_nonce) {
            return Err(Error::new(ErrorKind::Auth, "SCRAM server nonce mismatch"));
        }
        let salt_b64 = attrs
            .get("s")
            .ok_or_else(|| Error::new(ErrorKind::Auth, "SCRAM missing salt"))?;
        let iter = attrs
            .get("i")
            .ok_or_else(|| Error::new(ErrorKind::Auth, "SCRAM missing iterations"))?
            .parse::<u32>()
            .map_err(|_| Error::new(ErrorKind::Auth, "SCRAM invalid iterations"))?;
        let salt = general_purpose::STANDARD
            .decode(salt_b64.as_bytes())
            .map_err(|_| Error::new(ErrorKind::Auth, "SCRAM invalid salt"))?;

        let mut salted = [0u8; 32];
        pbkdf2_hmac::<Sha256>(password.as_bytes(), &salt, iter, &mut salted);
        let client_key = hmac_bytes(&salted, b"Client Key");
        let stored_key = sha256_bytes(&client_key);
        let client_final_without_proof = format!("c=biws,r={}", nonce);
        let auth_message = format!(
            "{},{},{}",
            self.client_first_bare, server_first, client_final_without_proof
        );
        let client_signature = hmac_bytes(&stored_key, auth_message.as_bytes());
        let client_proof = xor_bytes(&client_key, &client_signature);
        let server_key = hmac_bytes(&salted, b"Server Key");
        self.server_signature = Some(hmac_bytes(&server_key, auth_message.as_bytes()));
        let proof_b64 = general_purpose::STANDARD.encode(client_proof);
        Ok(format!("{},p={}", client_final_without_proof, proof_b64))
    }

    pub fn verify_server_final(&self, server_final: &str) -> Result<()> {
        let attrs = parse_attributes(server_final);
        let verifier = attrs
            .get("v")
            .ok_or_else(|| Error::new(ErrorKind::Auth, "SCRAM missing verifier"))?;
        let signature = self
            .server_signature
            .as_ref()
            .ok_or_else(|| Error::new(ErrorKind::Auth, "SCRAM missing server signature"))?;
        let expected = general_purpose::STANDARD.encode(signature);
        if &expected != verifier {
            return Err(Error::new(
                ErrorKind::Auth,
                "SCRAM server signature mismatch",
            ));
        }
        Ok(())
    }
}

fn parse_attributes(message: &str) -> std::collections::HashMap<String, String> {
    let mut attrs = std::collections::HashMap::new();
    for part in message.split(',') {
        if let Some(idx) = part.find('=') {
            let key = &part[..idx];
            let value = &part[idx + 1..];
            attrs.insert(key.to_string(), value.to_string());
        }
    }
    attrs
}

fn hmac_bytes(key: &[u8], data: &[u8]) -> Vec<u8> {
    let mut mac = HmacSha256::new_from_slice(key).expect("hmac key");
    mac.update(data);
    mac.finalize().into_bytes().to_vec()
}

fn sha256_bytes(data: &[u8]) -> Vec<u8> {
    use sha2::Digest;
    let mut hasher = Sha256::new();
    hasher.update(data);
    hasher.finalize().to_vec()
}

fn xor_bytes(left: &[u8], right: &[u8]) -> Vec<u8> {
    left.iter().zip(right.iter()).map(|(l, r)| l ^ r).collect()
}
